Fix error-prone SQL queries

ClearlyClaire opened the pull request:
Following the news of the compromise of one infamous Mastodon fork, I have been reviewing Mastodon's source code for SQL injection vulnerabilities.

So far, I have found none, but I have found a few queries that had me scratch my heard pretty hard and that could easily introduce vulnerabilities in the future if we are not careful.

This PR should make them more robust.

ClearlyClaire committed:
Fix error-prone SQL queries in Account search

While this code seems to not present an actual vulnerability, one could
easily be introduced by mistake due to how the query is built.

This PR parameterises the `to_tsquery` input to make the query more robust.

ClearlyClaire committed:
Harden code for Status#tagged_with_all and Status#tagged_with_none

Those two scopes aren't used in a way that could be vulnerable to an SQL
injection, but keeping them unchanged might be a hazard.

ClearlyClaire committed:
Remove unneeded spaces surrounding tsquery term

ClearlyClaire commented:
Removed unneeded space characters around the generated tsquery, I do not think they serve any purpose, I tried a bunch of tsqueries terms, and encoding them with or without spaces always result in the same tsquery.

ClearlyClaire commented:
I still think this is a worthy change.

mohe2015 commented on app/models/status.rb:
```diff
@@ -96,12 +96,12 @@ class Status < ApplicationRecord
   scope :not_domain_blocked_by_account, ->(account) { account.excluded_from_timeline_domains.blank? ? left_outer_joins(:account) : left_outer_joins(:account).where('accounts.domain IS NULL OR accounts.domain NOT IN (?)', account.excluded_from_timeline_domains) }
   scope :tagged_with_all, ->(tag_ids) {
-    Array(tag_ids).reduce(self) do |result, id|
+    Array(tag_ids).map(&:to_i).reduce(self) do |result, id|
```
I don't know ruby but the previous here looks quite dangerous as far as I can tell. Are the tag_ids sanitized somewhere else or is there another reason why this is not an sql injection?

Gargron replied:
The ids are not passed from user input

mohe2015 replied:
Still prepared statements are not possible when combining queries?

ClearlyClaire replied:
No. That's what `tagged_with` does (returning statuses that are tagged with any of the given tags). `tagged_with_all` needs to return  tagged with *all* the given tags, there might be better way to query that, but right now I don't know which, and `tag_id IN (:tag_ids)` isn't one of them. Though `tagged_with_none` can most probably be rewritten in a better way using `tag_id IN (:tag_ids)`.

ClearlyClaire replied:
I have rewritten `tagged_with_none` in a way that seems slightly more efficient and does not involve creating names from any input.

mohe2015 reviewed (approved):
I don't know ruby that much but this should *really* get merged. You need to rebase though.

ClearlyClaire committed:
Please CodeClimate

ClearlyClaire committed:
Move advanced_search_for SQL template to its own function

This avoids one level of indentation while making clearer that the SQL template
isn't build from all the dynamic parameters of advanced_search_for.

ClearlyClaire committed:
Add tests covering tagged_with, tagged_with_all and tagged_with_none

mohe2015 commented on app/models/account.rb:
```diff
@@ -431,98 +434,96 @@ def inboxes
+      records
+    end
+
+    def from_text(text)
```
the rebase doesn't seem to be perfect as this functoin got moved but doesn't seem to have changed.

ClearlyClaire replied:
It's not supposed to have changed. It's just noise from me having moved a bunch of code from `advanced_search_for` to a private method.

mohe2015 replied:
Yeah I know - just thought maybe the diff can be improved. But is not that important.

ClearlyClaire replied:
Since the relative position of the two chunks of code were swapped (because I moved something to a private method), the diff would necessarily represent one as moved, so there's no way to make the diff nicer but that's just an issue with the diff representation and not the code itself.

ClearlyClaire committed:
Rewrite tagged_with_none to avoid multiple joins and make it more robust

ClearlyClaire commented:
Since the relative position of the two chunks of code were swapped (because I moved something to a private method), the diff would necessarily represent one as moved, so there's no real way to make the diff nicer but that's just an issue with the diff representation and not the code itself.

ClearlyClaire committed:
Remove obsolete brakeman warnings

ClearlyClaire commented:
I still think this is a good change. Rebased.

Gargron commented on app/models/account.rb:
```diff
@@ -438,98 +441,96 @@ def inboxes
+      terms = unsanitized_terms.gsub(DISALLOWED_TSQUERY_CHARACTERS, ' ')
+
+      # The final ":*" is for prefix search.
+      "'#{terms}':*"
```
I think I mentioned this before: This does not yield the same results as `''' ' || #{terms} || ' ''' || ':*'`

ClearlyClaire replied:
Which different results? `''' ' || #{terms} || ' ''' || ':*'` (when `#{terms}` get substituted by a string surrounded by single-quotes and not containing some characters) is a way to write the string `' #{terms} ':*` in SQL (it's tricky because of the single-quotes), `"' #{terms} ':*"` results in the same string, which cannot be pasted directly in SQL, but it's ok because it's used as a parameter to a parameterized query instead.

So the difference between the two strings is just the spaces surrounding `#{terms}`. I understand the query is from https://github.com/Casecommons/pg_search/blob/master/lib/pg_search/features/tsearch.rb#L100-L138 but I don't understand what those extra spaces would mean, I could not understand it from PostgreSQL's documentation and did not observe any difference on testing data.

ClearlyClaire replied:
Since I could not understand the reason for the surrounding spaces in the built query, I ended up comparing the generated `tsqueries` (at the SQL level) using the following code:
```ruby
DISALLOWED_TSQUERY_CHARACTERS = /['?\\:‘’]/.freeze

def compare_tsqueries(unsanitized_terms)
  terms = unsanitized_terms.gsub(DISALLOWED_TSQUERY_CHARACTERS, ' ')
  inline_terms = Arel.sql(ActiveRecord::Base.connection.quote(terms))
  inline_tsquery = "to_tsquery('simple', ''' ' || #{inline_terms} || ' ''' || ':*')"
  tsquery_source = "'#{terms}':*"

  sql = "SELECT #{inline_tsquery} = to_tsquery('simple', $1)"
  ActiveRecord::Base.connection.select_all(sql, nil, [[nil, tsquery_source]]).rows[0][0]
end

def random_ascii_string(length)
  value = ''
  length.times do
    value  << (32 + rand(126-32)).chr
  end
  value
end

1000000.times.map { random_ascii_string(10) }.filter { |x| !compare_tsqueries(x) }
```

While I don't understand that result, it seems `to_tsquery('simple', query)` works slightly differently if the first term of query starts with `./`, `../` or `.. `. I don't know how to explain that change, so I'm introducing the spaces back, even though I do not think these discrepancies (one case searching for a leading `.` less) would cause any issue at all.

ClearlyClaire committed:
Revert "Remove unneeded spaces surrounding tsquery term"

The two queries are not strictly equivalent.

This reverts commit 86f16c537e06c6ba4a8b250f25dcce9f049023ff.
